src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -0,0 +1,180 @@
+      { (mb, args) =>
+        val cts = types.init.map(TypeToIRIntermediateClassTag(_).runtimeClass).toArray
+        Code.invokeScalaObject[R](obj, method, cts, args)
+      }
+  }
+
+  def registerJavaStaticFunction[C: ClassTag, R: ClassTag](mname: String, types: Type*)(method: String) {
+    registerCode[R](mname, types.toArray)
+      { (mb, args) =>
+        val cts = types.init.map(TypeToIRIntermediateClassTag(_).runtimeClass)
+        Code.invokeStatic[C, R](method, cts.toArray, args)
+      }
+  }
+
+  def registerIR(mname: String, types: Array[Type])(f: Seq[IR] => IR) {
+    addIR(mname, types.init, f)
+  }
+
+  def registerCode[R](mname: String, rt: Type)(impl: MethodBuilder => Code[R]): Unit =
+    registerCode[R](mname, Array(rt)) { case (mb, Array()) => impl(mb) }
+
+  def registerCode[R](mname: String, mt1: Type, rt: Type)(impl: (MethodBuilder, Code[_]) => Code[R]): Unit =
+    registerCode[R](mname, Array(mt1, rt)) { case (mb, Array(a1)) => impl(mb, a1) }
+
+  def registerCode[R](mname: String, mt1: Type, mt2: Type, rt: Type)(impl: (MethodBuilder, Code[_], Code[_]) => Code[R]): Unit =
+    registerCode[R](mname, Array(mt1, mt2, rt)) { case (mb, Array(a1, a2)) => impl(mb, a1, a2) }
+
+  def registerCode[R](mname: String, mt1: Type, mt2: Type, mt3: Type, rt: Type)(impl: (MethodBuilder, Code[_], Code[_], Code[_]) => Code[R]): Unit =
+    registerCode[R](mname, Array(mt1, mt2, mt3, rt)) { case (mb, Array(a1, a2, a3)) => impl(mb, a1, a2, a3) }
+
+  def registerIR(mname: String, rt: Type)(f: () => IR): Unit =
+    registerIR(mname, Array(rt)){ case Seq() => f() }
+
+  def registerIR(mname: String, mt1: Type, rt: Type)(f: IR => IR): Unit =
+    registerIR(mname, Array(mt1, rt)){ case Seq(a1) => f(a1) }
+
+  def registerIR(mname: String, mt1: Type, mt2: Type, rt: Type)(f: (IR, IR) => IR): Unit =
+    registerIR(mname, Array(mt1, mt2, rt)){ case Seq(a1, a2) => f(a1, a2) }
+
+  def registerIR(mname: String, mt1: Type, mt2: Type, mt3: Type, rt: Type)(f: (IR, IR, IR) => IR): Unit =
+    registerIR(mname, Array(mt1, mt2, mt3, rt)){ case Seq(a1, a2, a3) => f(a1, a2, a3) }
+
+}
+
+object IRFromAST {
+
+  private def tryPrimOpConversion(fn: String): IndexedSeq[IR] => Option[IR] =
+    flatLift {
+      case IndexedSeq(x) => for {
+        op <- UnaryOp.fromString.lift(fn)
+        t <- UnaryOp.returnTypeOption(op, x.typ)
+      } yield ApplyUnaryPrimOp(op, x, t)
+      case IndexedSeq(x, y) => for {
+        op <- BinaryOp.fromString.lift(fn)
+        t <- BinaryOp.returnTypeOption(op, x.typ, y.typ)
+      } yield ApplyBinaryPrimOp(op, x, y, t)
+    }
+
+  def lookup(name: String, args: IndexedSeq[IR], types: IndexedSeq[Type]): Option[IR] = {
+    (name, args) match {
+      case ("isDefined", IndexedSeq(x)) => Some(ApplyUnaryPrimOp(Bang(), IsNA(x)))
+      case ("orMissing", IndexedSeq(cond, x)) => Some(If(cond, x, NA(x.typ)))
+      case ("size", IndexedSeq(x)) if x.typ.isInstanceOf[TArray] => Some(ArrayLen(x))
```

Comment:
Now that we have a function registry, I don't know why we need an separate mechanism for these.  Can we just make them registered functions?